app/src/main/java/org/mozilla/fenix/share/SaveToPDFMiddleware.kt

```diff
@@ -0,0 +1,43 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.share
+
+import android.content.Context
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.EngineAction
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.lib.state.Action
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.ui.tabcounter.TabCounter
+import org.mozilla.fenix.R
+import org.mozilla.fenix.components.FenixSnackbar
+import org.mozilla.fenix.components.FenixSnackbar.Companion.LENGTH_LONG
+import org.mozilla.fenix.ext.getRootView
+
+/**
+ * [BrowserAction] middleware reacting in response to Save to PDF related [Action]s.
+ * @property context The [TabCounter] that will be setup with event handlers.
+ */
+class SaveToPDFMiddleware(
+    private val context: Context,
+) : Middleware<BrowserState, BrowserAction> {
+    override fun invoke(
+        ctx: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction,
+    ) {
+        if (action is EngineAction.SaveToPdfErrorAction) {
+            val rootView = context.getRootView() ?: return
```

Comment:
When testing, `context.getRootView()` was always `null`, so this code path was always early returned. This is probably because `context.isUiContext` is always `false` and when I queried for `context.display`, the error message was more useful:
```
java.lang.UnsupportedOperationException: Tried to obtain display from a Context not associated with one. Only visual Contexts (such as Activity or one created with Context#createWindowContext) or ones created with Context#createDisplayContext are associated with displays. Other types of Contexts are typically related to background entities and may return an arbitrary display.
```

Reply:
That said, if it did work however, I think we would be potentially opening up a memory leak here: if we using the _application_ Context to attach to a view, then that view is also going to be tied to the BrowserStore. If we do not safely release the `FenixSnackbar`, then I think we are going to hold onto some parts of the snackbar view for the application Context lifetime.

As for a solution to this potential problem, I think there are a few options:

 1. Dispatch some state change to the AppStore that will then be observed from a Context associated with a display to show the `Snackbar` and then consume the value.
 2. Similar to the above, but avoid the indirection and put this state straight into `BrowserStore`. Something a bit like `DownloadState` but on the root of the state object, I think? 
 3. (Not recommended) Use a new pattern that can register/unregister from the middleware so it can observe the action changes from a listener and call `FenixSnackbar` from an appropriate context (like HomeActivity). We're now re-creating how to observe state via odd loop-holes so we should not do this. It was more of a thought experiment. 🙂